### ISSUE TITLE
Can I use WSL on Windows 10 to configure Nginx with ModSecurity as reverse proxy?

### ISSUE THREAD

**Issue by shen1212** [SpiderLabs/ModSecurity#2926]
Hi everyone, I'm a cybersecurity degree student currently working on my localhost and I wish to apply ModSecurity as my security enhancements. I have done installing nginx and ModSecurity in Ubuntu WSL that is following a YouTube tutorial and some research. It is working on my machine but it some how does not trigger 403 when I access "http://192.168.0.11:3000/?exec=/bin/bash" but is able to trigger on http://127.0.0.1/?exec=/bin/bash.
I have added this configuration in modsecurity.conf to try to trigger 403 but that does not seems to work

Custom Rule -----------------------------------------------------------
Allow requests from 192.168.0.11
SecRule REMOTE_ADDR "@ipMatch 192.168.0.11" \
"id:'100001',phase:1,t:none,pass,nolog"
Deny requests from other IP addresses
SecRule REMOTE_ADDR "!@ipMatch 192.168.0.11" \
"id:'100002',phase:1,t:none,log,deny,msg:'Access denied from %{REMOTE_ADDR}',severity:2"

YouTube tutorial that I followed: https://www.youtube.com/watch?v=5eRxOYbaIEI
I'm sorry if I'm using Issues the wrong way but I'm desperate.. 
Can some one teach me how to do it...?

**Comment by martinhsv** [comment]
Hello @shen1212 ,

It sounds like probably the ModSecurity engine is not receiving the first request at all.

You can confirm this via the ModSecurity debug log (probably use level '9', but at minimum level '4'). If ModSecurity sees a transaction, you should at least see a log line that says 'Initializing transaction'.

If there is no output that might mean you have an issue in your nginx configuration that results in requests to port 80 being directed to ModSecurity, but requests to port 3000 are not.

**Comment by shen1212** [comment]
I have configured the debug log and I think that's the issue that you mentioned as it did not return any log when I try to trigger it.
SecDebugLog /var/log/modsec_debug.log
SecDebugLogLevel 9
But I have actually configured the nginx to listen to port 3000 and port 3001 as I have now implemented SSL on my website.
I have changed the port without the IP address and still....
Below is the configuration:
server{
        listen 443 ssl;
        listen 3000;
        listen 3001;
        server_name localhost;

        modsecurity on;
        modsecurity_rules_file /etc/nginx/modsec/main.conf;
        ssl_certificate /etc/nginx/sites-available/localhost.pem;
        ssl_certificate_key /etc/nginx/sites-available/localhost-key.pem;

        location / {
            proxy_pass http://localhost:3000;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection 'upgrade';
            proxy_set_header Host $host;
            proxy_cache_bypass $http_upgrade;
        }
        }
I'm sorry, am I configuring it wrong? Or did you need any other .conf files?

**Comment by martinhsv** [comment]
Hi @shen1212 ,

I made exactly 1 change to a very basic nginx setup (including no reverse proxy), and nginx processed the request and provided the request to ModSecurity. That one line right after the original line to listen on port 80:
```
server {
    listen       80;
    listen       3000;
...
```

I'm not clear on what your purpose is with using proxy_pass to the same host. You did reference reverse proxy in your title, but 
I have never previously set up nginx as an RP to self. One thing that seems surprising to me, is that within the same 'server' block you are both a) listening on port 3000 and b) redirecting traffic to localhost:3000. Isn't that likely to produce an attempt at an infinite loop? (At least that happened to me when I tried to mimic your setup ... nginx eventually failed with an error mesage '768 worker_connections are not enough'.) And yet you say it is working except for that ModSecurity never sees the transaction?!

This may be a case where a more esoteric nginx configuration needs comment from someone with more nginx expertise.

**Comment by shen1212** [comment]
Thank you for your reply and help.
I'm really sorry that I'm unclear with my question.
I need to develop a website with security enhancements, and I found out that nginx with ModSecurity can do quite a lot of thing, so I decided to use it as WAF for my web. But the problem is I need to use WSL to work on it since my machine is Windows OS. So I do some research and also chatgpt for the solution and this is the result of it. 
The problem exist when I started my localhost 3000 (npm run dev) in Windows machine as it seems to override or just ignore nginx. It did return 403 when I did not run my web for both yours and my setup without error on WSL site.
I asked ChatGPT Plus for this and the solution given did not work out but I still keep the config. Here is all of the files that are related and default.txt have 3 different configuration files and deleted most of the comments thought that might be easier to read... FYI I have also included custom rules that will trigger 403 no matter what but still it did not return 403 when my web's running.
I'm very thankful for you to help me out with this but I'd still be thankful if you had enough with this.

[default.txt](https://github.com/SpiderLabs/ModSecurity/files/12041951/default.txt)
[modsec rules.txt](https://github.com/SpiderLabs/ModSecurity/files/12041952/modsec.rules.txt)
[Nginx.conf.txt](https://github.com/SpiderLabs/ModSecurity/files/12041953/Nginx.conf.txt)

Thanks in advance!
